### PR TITLE
Set up Danger

### PR DESCRIPTION
**Pull request by rgarner**
- Add the danger gem, see http://danger.systems/ruby/ for details
- Add a Dangerfile to check for Changelog additions in a PR
- Make Circle run this in its test:overrides
- Separate to this, we have set up public access with a
  DANGER_GITHUB_API_TOKEN

Note, we didn't set up the "drastic" option for Circle to build from forks (seen in the interactive session below). What we've done *should* allow Circle to tell Github when Danger is warning us about something, at least when this PR is merged.

When this is merged we should subsequently try and make a small app change (maybe a copy change) without a CHANGELOG change. We should see that merging is blocked.

Here's the whole interactive shebang:

```
➜  consent-form-builder-rails git:(danger-danger---high-voltage) ✗ danger init

OK, thanks russ, have a seat and we'll get you started.

We need to do the following:
 - [ ] Create a Dangerfile and add a few simple rules.
 - [ ] Create a GitHub account for Danger to use, for messaging.
 - [ ] Set up an access token for Danger.
 - [ ] Set up Danger to run on your CI.

Step 1: Creating a starter Dangerfile

I've set up an example Dangerfile for you in this folder.
cat /Users/russ/dev/bdos/consent-form-builder-rails/Dangerfile

  # Sometimes it's a README fix, or something like that - which isn't relevant for
  # including in a project's CHANGELOG for example
  declared_trivial = github.pr_title.include? "#trivial"

  # Make it more obvious that a PR is a work in progress and shouldn't be merged yet
  warn("PR is classed as Work in Progress") if github.pr_title.include? "[WIP]"

  # Warn when there is a big PR
  warn("Big PR") if git.lines_of_code > 500

  # Don't let testing shortcuts get into master by accident
  fail("fdescribe left in tests") if `grep -r fdescribe specs/ `.length > 1
  fail("fit left in tests") if `grep -r fit specs/ `.length > 1

There's a collection of small, simple ideas in here, but Danger is about being able to easily
iterate. The power comes from you having the ability to codify fixes for some of the problems
that come up in day to day programming. It can be difficult to try and see those from day 1.

If you'd like to investigate the file, and make some changes - I'll wait here,
press return when you're ready to move on...


Step 2: Creating a GitHub account

In order to get the most out of Danger, I'd recommend giving her the ability to post in
the code-review comment section.

IMO, it's best to do this by using the private mode of your browser. Create an account like
Consent-form-builder-railsBot, and don't forget a cool robot avatar.

Here are great resources for creative commons images of robots:
 -> https://www.flickr.com/search/?text=robot&license=2%2C3%2C4%2C5%2C6%2C9
 -> https://www.google.com/search?q=robot&tbs=sur:fmc&tbm=isch&tbo=u&source=univ&sa=X&ved=0ahUKEwjgy8-f95jLAhWI7hoKHV_UD00QsAQIMQ&biw=1265&bih=1359
Consent-form-builder-railsBot will need access to your repo. Simply because the code is not available for the public
to read and comment on.

Note: Holding cmd ( ⌘ ) and clicking a link will open it in your browser.

Cool, please press return when you have your account ready (and you've verified the email...)


Step 3: Configuring a GitHub Personal Access Token

Here's the link, you should open this in the private session where you just created the new GitHub account
 -> https://github.com/settings/tokens/new

For token access rights, I need to know if this is for an Open Source or Closed Source project
? [ Open / Closed ]
> Open
For Open Source projects, I'd recommend giving the token the smallest scope possible.
This means only providing access to public_repo in the token.

This token limits Danger's abilities to just writing comments on OSS projects. I recommend
this because the token can quite easily be extracted from the environment via pull requests.

It is important that you do not store this token in your repository, as GitHub will automatically revoke it when pushed.

👍, please press return when you have your token set up...


Step 4: Add Danger for your CI

Add - bundle exec danger as a new step in the test:override: section of your circle.yml file.
As I'm not sure what CI you want to run Danger on based on the files in your repo, I'll just offer some generic
advice. You want to run bundle exec danger after your tests have finished running, it should still be during the testing
process so the build can fail.

OK, I'll give you a moment to do this...


Final step: exposing the GitHub token as an environment build variable.

As you have an Open Source repo, this token should be considered public, otherwise you cannot
run Danger on pull requests from forks, limiting its use.
Before we start, it's important to be up-front. CircleCI only really has one option to support running Danger
for forks on OSS repos. It is quite a drastic option, and I want to let you know the best place to understand
the ramifications of turning on a setting I'm about to advise.
 -> https://circleci.com/docs/fork-pr-builds
TLDR: If you have anything other than Danger config settings in CircleCI, then you should not turn on the setting.
I'll give you a minute to read it...


On danger/danger we turn on Permissive building of fork pull requests this exposes the token to Danger
You can find this setting at:
 -> https://circleci.com/gh/barnardos/consent-form-builder-rails/edit#advanced-settings

I'll hold...


In order to expose an environment variable, go to:
 -> https://circleci.com/gh/barnardos/consent-form-builder-rails/edit#env-vars
The name is DANGER_GITHUB_API_TOKEN and the value is the GitHub Personal Access Token.
You need to expose a token called DANGER_GITHUB_API_TOKEN and the value is the GitHub Personal Access Token.
Depending on the CI system, this may need to be done on the machine (in the ~/.bashprofile) or in a web UI somewhere.
We have a guide for all supported CI systems on danger.systems:
 -> http://danger.systems/guides/getting_started.html#setting-up-danger-to-run-on-your-ci
This is the last step, I can give you a second...


Useful info

- One of the best ways to test out new rules locally is via bundle exec danger pr.
- You can have Danger output all of her variables to the console via the --verbose option.
- You can look at the following Dangerfiles to get some more ideas:
 -> https://github.com/danger/danger/blob/master/Dangerfile
 -> https://github.com/artsy/eigen/blob/master/Dangerfile


🎉
And you're good to go. Danger is a collaboration between Orta Therox, Gem 'Danger' McShane and Felix Krause.
If you like Danger, let others know. If you want to know more, follow @orta and @KrauseFx on Twitter.
If you don't like Danger, help us improve the project! xxx
```

### CODE REVIEW

**Comment by rgarner** [comment]
Oh hello @barnardos-danger-bot, lovely to see you :robot:
